pkg/ddl/ddl_api.go

```diff
@@ -4626,14 +4626,15 @@ func (d *ddl) AlterTablePartitioning(ctx sessionctx.Context, ident ast.Ident, sp
 				return err
 			}
 			if !ck {
-				if ctx.GetSessionVars().EnableGlobalIndex {
-					return dbterror.ErrCancelledDDLJob.GenWithStack("global index is not supported yet for alter table partitioning")
-				}
-				indexTp := "UNIQUE INDEX"
-				if index.Primary {
-					indexTp = "PRIMARY"
+				if !ctx.GetSessionVars().EnableGlobalIndex || index.Primary {
```

Comment:
Short answer, we don't need to support it right now, but we should create a feature request for it.

Long answer:
My thoughts about Primary Key and Global key and if it is clustered or non-clustered.

So a primary key cannot be a global index, i.e. the primary key must include all partitioning columns.
Why? To ensure the uniqueness at table level, and, in my opinion, ensure that a primary key lookup is a single index lookup if clustered, and at most 2 lookups if non-clustered; pk index + heap (_tidb_rowid) 

If the PK is clustered and the table is partitioned, then the actual PK would need to be both clustered and include the partitioning column, to ensure a single key lookup.

So if the PK is non-clustered, could not the PK just be the Global Index?
I would say it could, but I don't think we need to support it yet, since one can accomplish the same with an unique index.

Non-Clustered index example, what we would like to have:
CREATE TABLE t (pk_col varchar(255), partition_col int, data_col JSON, primary key (pk_col) nonclustered) PARTITION BY HASH(partition_col) PARTITIONS 3
table/partition data:
Key: _tidb_rowid
Val: pk_col + partition_col + data_col
PK index:
Key: pk_col <= This is currently not supported
Val: partition_col + data_col
Global index, will be the same as the PK:
Key: pk_col
Val: partition_col + data_col

So what we can do is:
CREATE TABLE test.t (pk_col varchar(255) not null, partition_col int, data_col JSON, unique key (pk_col)) PARTITION BY HASH(partition_col) PARTITIONS 3;
Which will create a non-clustered table the way we want it (pk_col is really the PK, just not explicitly named as such!)
Table/partition heap/non-clustered table:
Key: _tidb_rowid
Val: pk_col + partition_col + data_col
Global, Unique Key:
Key: pk_col
Val: Partition ID + _tidb_rowid

But what is missing is to allow partition a table with a PRIMARY KEY, which could be allowed, if it was an non-clustered primary key.



For clustered primary key, what we would prefer to do is:
CREATE TABLE t (pk_col varchar(255), partition_col int, data_col JSON, primary key (pk_col) clustered) PARTITION BY HASH(partition_col) PARTITIONS 3
Clustered index example:
Key: pk_col <= This would not allow a single key lookup, since it would need the partition_col as well.
Val: partition_col + data_col
+ a global index:
Key: pk_col <= This would add the lookup of the Global index.
Val: partition_col
But it has to be:
CREATE TABLE t (pk_col varchar(255), partition_col int, data_col JSON, primary key (pk_col, partition_col) clustered, unique key (pk_col)) PARTITION BY HASH(partition_col) PARTITIONS 3;
Key: pk_col + partition_col
Val: data_col
+ a global index:
Key: pk_col
Val: partition_col (+ pk_col?)
Which means that tidb could enhance the optimizer to read from the actual primary key, skipping the lookup through the global index (only using it for writes, to ensure uniqueness).

What would be needed to further discuss is which decision TiDB should take and which decisions should be left to the user/admin, like:
- Adding a global index, resulting in Partition management commands like drop/truncate/exchange partition is no-longer near instant, but now would take some time to reorganize the global index.
- Adding a global index for a clustered table would really have both the partitioned clustered PK a global unique non-null index.
These questions are also connected to if we should display if an index is global, in SHOW CREATE TABLE, and also accept GLOBAL and possibly LOCAL as index options.

Reply:
```
mysql> create table t(a int, b int) partition by hash(b) partitions 5;
Query OK, 0 rows affected (0.05 sec)

mysql> alter table t add primary key (a);
Query OK, 0 rows affected (0.34 sec)

mysql> alter table t remove partitioning;
Query OK, 0 rows affected (0.20 sec)

mysql> alter table t partition by hash(b) partitions 5;
ERROR 1503 (HY000): A PRIMARY must include all columns in the table's partitioning function
```

Based on the SQLs above, I think we should let the behaviours reasonable.